arangod/Aql/AstNode.cpp

```diff
@@ -376,6 +383,57 @@ static bool IsEmptyString(char const* p, size_t length) {
   return true;
 }
 
+/*static*/ std::string AstNode::encodeDataSourceType(
+    char* const name, size_t size, AstNode::DataSourceType type
+) {
+  StringRef suffix;
+
+  switch (type) {
+    case AstNode::DataSourceType::View:
+      suffix = VIEW_NODE_SUFFIX;
+      break;
+    case AstNode::DataSourceType::Collection:
+      suffix = COLLECTION_NODE_SUFFIX;
+      break;
+    default: // shut up compiler
+      break;
+  }
+
+  std::string param(name, size + suffix.size());
+  std::memcpy(&param[0] + size, suffix.data(), suffix.size());
```

Comment:
A string's underlying data is not supposed to be modified from the outside.

Reply:
well it's debatable. In IResearch we widely use std::string as a buffer. But because of the previous issue I'm going to use `reserve + append` there.

Reply:
I think technically this has changed in c++11, because an std::string now guarantees to store its data in contiguous memory. And if the string is non-empty (which is implicitly guaranteed by the callers of this function, but not by the function itself), it should work.